pkg/spoke/controllers/manifestwork_controller.go

```diff
@@ -2,46 +2,305 @@ package controllers
-func (m *ManifestWorkController) Run(ctx context.Context, recorder events.Recorder) {
-	factory.New().
-		WithInformers(m.manifestWorkInformer.Informer()).
-		WithSync(m.sync).ResyncEvery(5*time.Minute).ToController("ManifestWorkAgent", recorder).Run(ctx, 1)
+func (m *ManifestWorkController) applyManifest(manifests []runtime.RawExtension, recorder events.Recorder) []resourceapply.ApplyResult {
+	clientHolder := resourceapply.NewClientHolder().
+		WithAPIExtensionsClient(m.spokeAPIExtensionClient).
+		WithKubernetes(m.spokeKubeclient)
+
+	// Using index as the file name and apply manifests
+	indexstrings := []string{}
+	for i := 0; i < len(manifests); i++ {
+		indexstrings = append(indexstrings, strconv.Itoa(i))
+	}
+	results := resourceapply.ApplyDirectly(clientHolder, recorder, func(name string) ([]byte, error) {
+		index, _ := strconv.ParseInt(name, 10, 32)
+		return manifests[index].Raw, nil
+	}, indexstrings...)
+
+	// Try apply with dynamic client if the manifest cannot be decoded by scheme or typed client is not found
+	// TODO we should check the certain error.
+	for index, result := range results {
+		// Use dynamic client when scheme cannot decode manifest or typed client cannot handle the object
+		if isDecodeError(result.Error) || isUnhandledError(result.Error) {
+			results[index].Result, results[index].Changed, results[index].Error = m.applyUnstructrued(manifests[index].Raw, recorder)
+		}
+	}
+	return results
+}
+
+func (m *ManifestWorkController) decodeUnstructured(data []byte) (schema.GroupVersionResource, *unstructured.Unstructured, error) {
+	unstructuredObj := &unstructured.Unstructured{}
+	err := unstructuredObj.UnmarshalJSON(data)
+	if err != nil {
+		return schema.GroupVersionResource{}, nil, fmt.Errorf("Failed to decode object: %w", err)
+	}
+	mapping, err := m.restMapper.MappingForGVK(unstructuredObj.GroupVersionKind())
+	if err != nil {
+		return schema.GroupVersionResource{}, nil, fmt.Errorf("Failed to find grv from restmapping: %w", err)
+	}
+
+	return mapping.Resource, unstructuredObj, nil
+}
+
+func (m *ManifestWorkController) applyUnstructrued(data []byte, recorder events.Recorder) (*unstructured.Unstructured, bool, error) {
+	gvr, required, err := m.decodeUnstructured(data)
+	if err != nil {
+		return nil, false, err
+	}
+
+	existing, err := m.spokeDynamicClient.
+		Resource(gvr).
+		Namespace(required.GetNamespace()).
+		Get(context.TODO(), required.GetName(), metav1.GetOptions{})
+
+	if errors.IsNotFound(err) {
+		actual, err := m.spokeDynamicClient.Resource(gvr).Namespace(required.GetNamespace()).Create(
+			context.TODO(), required, metav1.CreateOptions{})
+		recorder.Eventf(fmt.Sprintf(
+			"%s Created", required.GetKind()), "Created %s/%s because it was missing", required.GetNamespace(), required.GetName())
+		return actual, true, err
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Do not update unstructured type right now. Need to figure out how to merge.
```

Comment:
i am thinking maybe we need to check whether an update is needed, by comparing desired and existing. For instance, try generate a patch and if patch is none, do not update. Otherwise the resource might be updated in every resync.